### PR TITLE
Add Gradient SHAP for layer and neuron

### PR DESCRIPTION
**Pull request by NarineK**
This PR adds Gradient SHAP for layer and neuron with documentation and tests cases including tests for data parallel.
It also moves `_compute_conv_delta_and_format_attrs ` function to `common.py` in order to be able to use it both in DeepLift and Gradient SHAP.
Some of the auxiliary test functions  are also shared between Gradient SHAP and DeepLift now.

### CODE REVIEW

File: tests/attr/test_data_parallel.py [hunk 1]
```diff
@@ -454,6 +508,7 @@ def _data_parallel_test_assert(
                     )
                 else:
                     attributions_orig = attr_orig.attribute(**kwargs)
+            self.setUp()
```
Comment (NarineK) [inline tests/attr/test_data_parallel.py]: I added this for determinism - in order to reinitiate the seeds. I'll pass seed to attribute in a separate PR.

File: captum/attr/_core/layer/layer_gradient_shap.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    ):
+        rand_coefficient = torch.tensor(
+            np.random.uniform(0.0, 1.0, inputs[0].shape[0]),
+            device=inputs[0].device,
+            dtype=inputs[0].dtype,
+        )
+
+        input_baseline_scaled = tuple(
+            self._scale_input(input, baseline, rand_coefficient)
+            for input, baseline in zip(inputs, baselines)
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/layer/layer_gradient_shap.py]: Currently, this class is not exposed. I was thinking to expose it, but that can be also done in a separate PR.

File: tests/attr/layer/test_layer_gradient_shap.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    def test_classification(self):
+        def custom_baseline_fn(inputs):
+            num_in = inputs[0].shape[1]
+            return (torch.arange(0.0, num_in * 4.0).reshape(4, num_in),)
+
+        num_in = 40
+        n_samples = 10
+
+        # 10-class classification model
+        model = SoftmaxModel(num_in, 20, 10)
```
[the remark this answers is not in the document]
Comment (NarineK) [reply tests/attr/layer/test_layer_gradient_shap.py]: This class is used in many places. We can set fixed weights in a separate PR for `SoftmaxModel `
